src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -118,12 +118,15 @@ ClientImpl::ClientImpl(ClientConfig config)
     , m_roundtrip_time_handler{std::move(config.roundtrip_time_handler)}
     , m_user_agent_string{make_user_agent_string(config)} // Throws
     , m_service{}                                         // Throws
```

Comment:
I think the `Service` here is a knot we have yet to untie. Right now, the service being owned by the `ClientImpl` lets us schedule async I/O and perform all of the logic related to the sync client in a single thread. Things like timeouts, retrying errors, and queuing download/upload completion notifications all happen sequentially with respect to the `Service` and all happen on a single thread along-side I/O. I think we'll need to have a separate websocket implementation done to see if it's okay to have the thread doing I/O be different than the thread doing sync client work. 

cc @RedBeard0531 , who did the original untying of this knot to see if he's already done any testing along these lines.

Reply:
I think it is OK for the I/O thread to be separate from the sync client thread, but in addition to never testing that, there are a few complications:

# the networking callbacks would need to either run on the sync client thread, or schedule themselves to the callback thread immediately
  # I think we also need to ensure that we either run the destructors for the callback on the sync thread or at least ensure it isn't running in parallel. Or we would need to check every callback to ensure that it doesn't capture anything by value that isn't safe to destroy on another thread.
# Our networking lib has some "interesting" destruction semantics. There is a comment about how code can rely on the fact that after destroying an object, its registered callback will never be called because the destructor and the callback would run on the same thread. I don't know how much the code is relying on this. I think it would be possible to preserve that semantic, but it will require some care to ensure it. And it may be better to try to change the code to not rely on it anyway.

Reply:
To be clear, I think it is a good idea to remove any implicit assumptions about single-threadedness from this code and explicitly use either mutexes or strands where needed. But I think that the path of least resistence would be to maintain the existing single-threaded behavior for now, then loosen the requirements separately.

Reply:
For the purposes of shipping this PR though, should we implement a strand to ensure sync client callbacks execute in the sync client thread as they always have? Right now all the callbacks from I/O have basically no locks or sequencing at all. They rely on our networking library's implementation to schedule things in the right order on a single thread.

Reply:
So I think at minimum, the `SocketFactory` will need to implement [`void post(Handler&&)`](https://github.com/realm/realm-core/blob/cbab9449a284b09a6c6b60c7c9d6acaa5e556270/src/realm/util/network.hpp#L348) method from `Service` and calls to `m_service.post()` will need to call that instead.